### ISSUE TITLE
Duplicate entries in junction tables

### ISSUE THREAD

**Issue by pedro3900** [balderdashy/sails#3570]
Hi,

I´am using sailsjs, and I have this problem:

I created 2 model related with a many to many relation, there are any posibility to have duplicated entries in this junction table, example:
Suposed that we have 2 tables 'a' and 'b' related.
juntion table
id  a_b__b_a    b_a__a_b
1          1                 1
2          1                 2
3          1                 2
4          1                 2
5          1                 2
6          1                 2
7          1                 2
8          1                 2
9          1                 2

I receive this error:

err: [Error: Trying to '.add()' an instance which already exists!]

**Comment by sgress454** [comment]
Hi @pedro3900, thanks for using Sails.  Please read our [guidelines for posting issues](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md#ii-issue-contributions) and update your post accordingly so we can better help.  In this case, searching the repo for "many-to-many duplicate" reveals at least one [existing issue](https://github.com/balderdashy/sails/issues/1647) that might provide some insight for you.  There's a good chance that you are experiencing a race condition--Sails doesn't implement transactions for you (yet) and you have to be careful to write asynchronous code correctly.  If none of that helps, we'll need more information before we can dig into this!

**Comment by mikermcneil** [comment]
@pedro3900 I'm going to go ahead and close this since it appears to be a duplicate of https://github.com/balderdashy/sails/issues/1647.  If you find that we're wrong and that's not the case, please create a GitHub repo with a Sails app that reproduces this and open a new issue-- we'll take a look asap.

**Comment by pedro3900** [comment]
The problem is not that sailsjs duplicate my entries when I save one entry, the problem is the oposite, I want to save duplicate entries with same value (except id field obviously that is autoinc) in the junction table. And when a save this, sails return me, err: [Error: Trying to '.add()' an instance which already exists!]

**Comment by luolongzhi** [comment]
I also face a wired problem is that there is duplicate records in my junction table, i don't know how this happen, but it really exist, like
id a_b__b_a b_a__a_b
1 1 1
2 1 2
3 1 2

id 2 3 are same, it is rare happen, noramlly it will return error when add same record.
 I think the problem is that:
1. 2 same request very fast, the first one is add save , the second is also add save
2. when first one is not complete save, just add complete; my second request is also add, then 
    at this time, 2 same record add, when save will success, so duplicate record happen.


After I face this problem, I check the mysql key schema property, I think in junction table, the 2 column
except the id should be union unique key, and this wire problem will avoid on the Database level;
So I do experiment, and will when create talbe let the UNIQUE KEY([col1, col2], 'xx'), so if on the code
level if can not avoid the duplicate record add, we can use the mysql dabase unique index to avoid


So, the question is:
if use mysql database, can when create table let the 2 column is union unique key, this can avoid duplicate record in some rarely situation

**Comment by irlnathan** [comment]
@pedro3900 @luolongzhi 

Howdy. If this is an issue, please check out the new [contribution guidelines](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md#ii-issue-contributions) and resubmit this in a new issue.  If this is more of a question, StackOverflow would be a better platform.

Best, Irl
